File: public/dsa/arrays/cumulative_sum/optimal_2/cumulative_sum.cpp

#include <bits/stdc++.h>
using namespace std;

vector<int> getCumulativeSum(vector<int> &arr) {
    int n = arr.size();
    vector<int> result(n);
    result[0] = arr[0];
    for (int i = 1; i < n; i++) {
        result[i] = result[i - 1] + arr[i];
    }
    return result;
}

// tests-start
int main() {
    vector<int> nums = {1, 2, 3, 4};
    auto result = getCumulativeSum(nums);
    for (int v : result) {
        cout << v << " ";
    }
    cout << endl;
    return 0;
}
// tests-end

/*output
1 3 6 10
*/

File: public/dsa/arrays/cumulative_sum/optimal_2/cumulative_sum.py

from typing import List


class Solution:
    def getCumulativeSum(self, arr: List[int]) -> List[int]:
        n = len(arr)
        result = [0 for _ in range(n)]
        result[0] = arr[0]
        for i in range(1, n):
            result[i] = result[i-1] + arr[i]
        return result


# tests-start
arr = [1, 2, 3, 4]
sol = Solution()
print(sol.getCumulativeSum(arr))
# tests-end

'''output
[1, 3, 6, 10]
'''
